internal/config_loader/types.go

```diff
@@ -118,49 +128,39 @@ type Parameter struct {
 // - Use Build for inline payload definitions directly in the config
 // - Use BuildRef to reference an external YAML file containing the build definition
 type Payload struct {
-	Name string `yaml:"name"`
+	Name string `yaml:"name" validate:"required"`
 	// Build contains a structure that will be evaluated and converted to JSON at runtime.
 	// The structure is kept as raw interface{} to allow flexible schema definitions.
 	// Mutually exclusive with BuildRef.
-	Build interface{} `yaml:"build,omitempty"`
+	Build interface{} `yaml:"build,omitempty" validate:"required_without=BuildRef,excluded_with=BuildRef"`
 	// BuildRef references an external YAML file containing the build definition.
 	// Mutually exclusive with Build.
-	BuildRef string `yaml:"buildRef,omitempty"`
+	BuildRef string `yaml:"buildRef,omitempty" validate:"required_without=Build,excluded_with=Build"`
 	// BuildRefContent holds the loaded content from BuildRef file (populated by loader)
 	BuildRefContent map[string]interface{} `yaml:"-"`
 }
 
-// Validate checks that the Payload configuration is valid.
-// Returns an error if:
-// - Both Build and BuildRef are set (mutually exclusive)
-// - Neither Build nor BuildRef is set (one is required)
-func (p *Payload) Validate() error {
-	hasBuild := p.Build != nil
-	hasBuildRef := p.BuildRef != ""
-
-	if hasBuild && hasBuildRef {
-		return fmt.Errorf("build and buildRef are mutually exclusive; set only one")
-	}
-	if !hasBuild && !hasBuildRef {
-		return fmt.Errorf("either build or buildRef must be set")
-	}
-	return nil
+// ActionBase contains common fields for action-like configurations.
+// Used by Precondition and PostAction to reduce duplication.
+type ActionBase struct {
+	Name    string     `yaml:"name" validate:"required"`
+	APICall *APICall   `yaml:"apiCall,omitempty" validate:"omitempty"`
+	Log     *LogAction `yaml:"log,omitempty"`
 }
 
-// Precondition represents a precondition check
+// Precondition represents a precondition check.
+// Must have at least one of: APICall (from ActionBase), Expression, or Conditions.
 type Precondition struct {
-	Name       string         `yaml:"name"`
-	APICall    *APICall       `yaml:"apiCall,omitempty"`
-	Capture    []CaptureField `yaml:"capture,omitempty"`
-	Conditions []Condition    `yaml:"conditions,omitempty"`
-	Expression string         `yaml:"expression,omitempty"`
-	Log        *LogAction     `yaml:"log,omitempty"`
+	ActionBase `yaml:",inline"`
+	Capture    []CaptureField `yaml:"capture,omitempty" validate:"dive"`
+	Conditions []Condition    `yaml:"conditions,omitempty" validate:"dive,required_without_all=ActionBase.APICall Expression"`
```

Comment:
Have you tested if this validation works correctly with the embedded field syntax?

The go-playground/validator may not support the ActionBase.APICall syntax for referencing embedded struct fields. The validator uses Go field names directly, not nested paths.